binary-transcoders/src/main/kotlin/com/kamelia/sprinkler/codec/binary/decoder/core/PrefixedArityReductionDecoder.kt

```diff
@@ -0,0 +1,98 @@
+package com.kamelia.sprinkler.codec.binary.decoder.core
+
+import com.kamelia.sprinkler.util.accumulate
+import com.kamelia.sprinkler.util.finish
+import com.kamelia.sprinkler.util.supply
+import java.util.stream.Collector
+
+/**
+ * A [Decoder] that decodes an object composed of a variable number of elements, by performing a
+ * [variable polyadic (aka variadic)](https://en.wikipedia.org/wiki/Arity#Varying_arity) reduction on a sequence of
+ * these elements.
+ *
+ * The element's byte sequence is prefixed by the number of elements to decode. It accumulates them until they are all
+ * decoded, and finally creates the resulting object and returns it.
+ *
+ * This decoder works by using a [Collector] for the accumulation and creation of the resulting object. It effectively
+ * does a reduction of the elements decoded by the [elementDecoder] into the resulting object.
+ *
+ * Here is an example of how to use this class:
+ *
+ * ```kotlin
+ * fun decode(byteDecoder: Decoder<Byte>) {
+ *     val decoder = PrefixedArityReductionDecoder(
+ *         Collectors.toList(),
+ *         byteDecoder,
+ *         byteDecoder
```

Comment:
just to be easier to read

```suggestion
 *         collector = Collectors.toList(),
 *         elementDecoder = byteDecoder,
 *         sizeDecoder = byteDecoder
```